src/core/drive/morph_renderer.js

```diff
@@ -80,35 +61,42 @@ export class MorphRenderer extends PageRenderer {
     }
   }
 
-  #shouldRemoveElement = (node) => {
-    return this.#shouldMorphElement(node)
+  static shouldRemoveElement = (node) => {
+    return this.shouldMorphElement(node)
   }
 
-  #reloadRemoteFrames() {
-    this.#remoteFrames().forEach((frame) => {
-      if (this.#isFrameReloadedWithMorph(frame)) {
-        this.#renderFrameWithMorph(frame)
-        frame.reload()
-      }
-    })
+  static isFrameReloadedWithMorph(element) {
+    return element.src && element.refresh === "morph"
   }
 
-  #renderFrameWithMorph(frame) {
-    frame.addEventListener("turbo:before-frame-render", (event) => {
-      event.detail.render = this.#morphFrameUpdate
-    }, { once: true })
+  // Public
+
+  async render() {
+    if (this.willRender) await this.#morphBody()
   }
 
-  #morphFrameUpdate = (currentElement, newElement) => {
-    dispatch("turbo:before-frame-morph", {
-      target: currentElement,
-      detail: { currentElement, newElement }
+  get renderMethod() {
+    return "morph"
+  }
+
+  // Private
+
+  async #morphBody() {
+    MorphRenderer.morphElements(this.currentElement, this.newElement)
+    this.#reloadRemoteFrames()
+
+    dispatch("turbo:morph", {
+      detail: {
+        currentElement: this.currentElement,
+        newElement: this.newElement
+      }
     })
-    this.#morphElements(currentElement, newElement.children, "innerHTML")
   }
 
-  #isFrameReloadedWithMorph(element) {
-    return element.src && element.refresh === "morph"
+  #reloadRemoteFrames() {
+    this.#remoteFrames().forEach((frame) => {
+      frame.reload()
+    })
```

Comment:
This class no longer needs to check if the frame has morph set or not. It can blindly trigger a `reload()` on each frame and trust that the frame will be smart — it knows if it needs to morph itself.